### ISSUE TITLE
How to disable log of libhv?

### ISSUE THREAD

**Issue by WorstCodeWay** [ithewei/libhv#171]
Hi, I'm using this great project in my project. Everything is perfect, but I want to disable the log from this project.

I mean close all logs totally, including console and log files. How to do that?

Thanks.

**Comment by ithewei** [comment]
hlog_disable

**Comment by WorstCodeWay** [comment]
> hlog_disable

I tried something very likely by calling `log_set_level(LOG_LEVEL_SILENT)`, but that doesn't work. I don't know if it is caused by the multiple instances of `http_server_t` and `hv::TcpServer`?

Besides, the log records the http server starting informaiton.

**Comment by ithewei** [comment]
You should call hio_disable before first log print, such as main() entrance.

**Comment by WorstCodeWay** [comment]
Hi, sorry for re-opening this thread. I find new problems here.
1. I cannot find function or macro called `hio_disable `
2. There are hv logs named `libhv.20230814.log` even though I have called `hlog_disable();` or `hlog_set_level(LOG_LEVEL_SILENT);` at the very beginning within `main()` and Its' contents are:
```
2023-08-14 02:36:10.502 INFO  EventLoopThread started, tid=22742 [EventLoopThread.h:90:loop_thread]
2023-08-14 02:36:10.604 INFO  EventLoopThread started, tid=22746 [EventLoopThread.h:90:loop_thread]
2023-08-14 09:30:35.746 INFO  EventLoopThread started, tid=4946 [EventLoopThread.h:90:loop_thread]
2023-08-14 09:30:35.850 INFO  EventLoopThread started, tid=4949 [EventLoopThread.h:90:loop_thread]
2023-08-14 09:38:44.332 INFO  EventLoopThread started, tid=5259 [EventLoopThread.h:90:loop_thread]
2023-08-14 09:38:44.434 INFO  EventLoopThread started, tid=5262 [EventLoopThread.h:90:loop_thread]
2023-08-14 09:38:59.485 INFO  EventLoopThread stopped, tid=5259 [EventLoopThread.h:109:loop_thread]
2023-08-14 09:38:59.486 INFO  EventLoopThread stopped, tid=5262 [EventLoopThread.h:109:loop_thread]
2023-08-14 09:45:29.054 INFO  EventLoopThread started, tid=5359 [EventLoopThread.h:90:loop_thread]
2023-08-14 09:45:29.159 INFO  EventLoopThread started, tid=5363 [EventLoopThread.h:90:loop_thread]
2023-08-14 09:47:07.061 INFO  EventLoopThread stopped, tid=5506 [EventLoopThread.h:109:loop_thread]
2023-08-14 09:47:07.062 INFO  EventLoopThread stopped, tid=5509 [EventLoopThread.h:109:loop_thread]
2023-08-14 09:53:36.331 INFO  EventLoopThread stopped, tid=5856 [EventLoopThread.h:109:loop_thread]
2023-08-14 09:53:36.332 INFO  EventLoopThread stopped, tid=5859 [EventLoopThread.h:109:loop_thread]
``` 
and it seems refreshed every time after my application closed.

**Comment by ithewei** [comment]
Test like this:
```
#include "hv/hlog.h"

int main() {
    hlogi("before");
    hlog_disable();
    hlogi("after");
    return 0;
}
```

**Comment by WorstCodeWay** [comment]
> Test like this:
> 
> ```
> #include "hv/hlog.h"
> 
> int main() {
>     hlogi("before");
>     hlog_disable();
>     hlogi("after");
>     return 0;
> }
> ```

@ithewei  Sorry for this seems no help. There is still a log file with following content:
```
2023-08-15 05:20:40.835 INFO  EventLoopThread stopped, tid=2335 [EventLoopThread.h:109:loop_thread]
2023-08-15 05:20:40.837 INFO  EventLoopThread stopped, tid=2338 [EventLoopThread.h:109:loop_thread]
```
And I find this log is recorded only after application closed.

**Comment by ithewei** [comment]
What I mean is that if you test this way, the correct result should be that before is printed but after is not printed.

**Comment by ithewei** [comment]
Maybe you should see if calling `hlog_set_level` elsewhere in your program changes the log level

**Comment by WorstCodeWay** [comment]
@ithewei  Ok, now I can make some conclusions. But we begin from these codes where logs sneaked from:
```
void loop_thread(const Functor& pre, const Functor& post) {
        hlogi("EventLoopThread started, tid=%ld", hv_gettid());
        setStatus(kStarted);

        if (pre) {
            loop_->queueInLoop([this, pre]{
                if (pre() != 0) {
                    loop_->stop();
                }
            });
        }

        loop_->run();
        assert(loop_->isStopped());

        if (post) {
            post();
        }

        setStatus(kStopped);
        hlogi("EventLoopThread stopped, tid=%ld", hv_gettid());
    }
```
1. By adding `hlog_disable();` into the very beginning of `main` can prevent `EventLoopThread started ...` logs into default log file
2. By adding `hlog_disable();` into the very beginning of `TcpClient::onConnection` callback can prevent `EventLoopThread stopped ...` logs into default log file
3. Forgetting any codes of previous steps will case logs sneaked into log file respectively.

So, I add `hlog_disable()` into two places to prevent all logs from `hvlib` itself,
However, I don't know why? Maybe, there is a bug ?

**Comment by ithewei** [comment]
No bug！
Do you use other libraries in your application, is it possible that you use other libraries, which encapsulate libhv, and call hlog_set_level?

**Comment by ithewei** [comment]
Another way is to set the handler to change the log output mode，like this：
```
void custom_log_handler(int loglevel, const char* buf, int len) {
    // do nothing
}

hlog_set_handler(custom_log_handler);
```

**Comment by WorstCodeWay** [comment]
@ithewei 
> No bug！ Do you use other libraries in your application, is it possible that you use other libraries, which encapsulate libhv, and call hlog_set_level?

Not really. My third-party dependencies are `qt`, `spdlog`, `opencv`, `pcl`, `hvlib`, that's all.

> Another way is to set the handler to change the log output mode，like this：
It's more clear, I'll try this.

**Comment by WorstCodeWay** [comment]
@ithewei 
> Another way is to set the handler to change the log output mode，like this：
> 
> ```
> void custom_log_handler(int loglevel, const char* buf, int len) {
>     // do nothing
> }
> 
> hlog_set_handler(custom_log_handler);
> ```

Unlucky, no helps either when putting it into beginning of `main`

**Comment by WorstCodeWay** [comment]
And new found! Every time there is a new `TcpClient`, I should add `hlog_disable` at beginning of `onConnection` callback to prevent logs.

**Comment by ithewei** [comment]
> @ithewei
> 
> > Another way is to set the handler to change the log output mode，like this：
> > ```
> > void custom_log_handler(int loglevel, const char* buf, int len) {
> >     // do nothing
> > }
> > 
> > hlog_set_handler(custom_log_handler);
> > ```
> 
> Unlucky, no helps either when putting it into beginning of `main`

This is impossible, unless you call hlog_destory halfway, are you sure you haven't called hlog_set_level and hlog_destory in your program?

**Comment by WorstCodeWay** [comment]
> > @ithewei
> > > Another way is to set the handler to change the log output mode，like this：
> > > ```
> > > void custom_log_handler(int loglevel, const char* buf, int len) {
> > >     // do nothing
> > > }
> > > 
> > > hlog_set_handler(custom_log_handler);
> > > ```
> > 
> > 
> > Unlucky, no helps either when putting it into beginning of `main`
> 
> This is impossible, unless you call hlog_destory halfway, are you sure you haven't called hlog_set_level and hlog_destory in your program?

I can confirm it by search all matches of `hlog_set_level` and `hlog_destory` in my codes, but find none.

### Please NOTE FOLLOWING
Maybe this should be helpful from hvlib source code in file hlog.c:
```
logger_t* hv_default_logger() {
    if (s_logger == NULL) {
        s_logger = logger_create();
        atexit(hv_destroy_default_logger);
    }
    return s_logger;
}
```
Here, `atexit` is interesting. 
> /* Register a function to be called when `exit' is called.  */
> extern int atexit (void (*__func) (void)) __THROW __nonnull ((1));

So, when there is a tcp-client connection, and after application(Qt application) is terminated. Can the client connected thread exit after than the main thread? If so, then the main thread will call `exit` (I don't know)? and then the client thread call `hlogi`, another `creation`!

**Comment by ithewei** [comment]
1、You can explicitly delete TcpClient；

**Comment by WorstCodeWay** [comment]
> 1、You can explicitly delete TcpClient before app exit； 2、remove the code line `hlogi("EventLoopThread stopped...` directly.

Yes, it's better.

**Comment by lovelypp777** [comment]
I encounted this problem too

**Comment by ithewei** [comment]
`libhv` 提供了关闭日志的接口，已支持：

```c
hlog_disable();            // 关闭所有日志（等价于把日志级别设为 LOG_LEVEL_SILENT）
```

定义见 `base/hlog.h`：
```c
#define hlog_disable()  logger_set_level(hlog, LOG_LEVEL_SILENT)
```

如果只想关闭文件同步/落盘，可用 `hlog_disable_fsync()`；也可以用 `logger_set_level()` 精细控制级别。

问题已解决，关闭此 issue。
